### PR TITLE
Enable pad_sequences to operate on sequences of numpy arrays

### PR DESCRIPTION
**Pull request by ddsh**
Hello,

As I am working  with LSTM using w2v embeddings, enabling pad_sequences to operate on sequence of numpy arrays would be very useful for such applications!

Robert

### CODE REVIEW

**Comment by fchollet** [comment]
What's your use case? 

From the documentation:

> `pad_sequences`
> Transform a list of `nb_samples` sequences (lists of scalars) into a 2D numpy array of shape `(nb_samples, nb_timesteps)`.

If what you're doing is significantly different from that, then rolling out another function would seem reasonable.

**Comment by ddsh** [comment]
The difference is in the embedding you chose for the word representations which are continuous np arrays.

To use the same terminology than in the documentation:
pad_sequences transforms a list of nb_samples sequences (list of scalars/ list of numpy arrays) into a 2D/3D numpy array of shape  (nb_samples, nb_timesteps)/(nb_samples, nb_timesteps, nb_dimension_embeddings)

File: keras/preprocessing/sequence.py [hunk 1]
```diff
@@ -4,31 +4,39 @@
 import random
 from six.moves import range
 
-def pad_sequences(sequences, maxlen=None, dtype='int32', padding='pre'):
+
+def pad_sequences(sequences, maxlen=None, dtype='int32', padding='pre', vector_sequence=False):
```
Comment (fchollet) [inline keras/preprocessing/sequence.py]: `vector_sequence=False` does not seem necessary, you could just check whether multidimensional arrays are being passed in `sequences`.

File: keras/preprocessing/sequence.py [hunk 2]
```diff
@@ -4,31 +4,39 @@
...
     nb_samples = len(sequences)
     if maxlen is None:
         maxlen = np.max(lengths)
-
-    x = np.zeros((nb_samples, maxlen)).astype(dtype)
+    
+    if not vector_sequence:
+        x = np.zeros((nb_samples, maxlen)).astype(dtype)
+    else:
+        dim = len(sequences[0][0])
```
Comment (fchollet) [inline keras/preprocessing/sequence.py]: What about the edge case where sequences is empty, or (more likely) a list where the first array is empty?

File: keras/preprocessing/sequence.py [hunk 3]
```diff
@@ -4,31 +4,39 @@
...
     for idx, s in enumerate(sequences):
         if padding == 'post':
             x[idx, :lengths[idx]] = s[:maxlen]
         else:
             x[idx, -min(maxlen, lengths[idx]):] = s[:maxlen]
     return x
-
-
+        
+
```
Comment (fchollet) [inline keras/preprocessing/sequence.py]: No such whitespace please.

**Comment by ddsh** [comment]
I made the changes accordingly to your comments.
(btw in the current version, an empty sequence provides an error)

File: keras/preprocessing/sequence.py [hunk 4]
```diff
@@ -4,24 +4,41 @@
...
     """
     lengths = [len(s) for s in sequences]
-
     nb_samples = len(sequences)
     if maxlen is None:
         maxlen = np.max(lengths)
-
-    x = np.zeros((nb_samples, maxlen)).astype(dtype)
+    
+    array_like = False
```
Comment (fchollet) [inline keras/preprocessing/sequence.py]: What about something like:
```python
try:
   array_like = isinstance(sequences[0][0], np.ndarray)
   dim = len(sequences[0][0])
except:
   raise Exception('"sequences" input to preprocessing.sequence.pad_sequences should be a non-empty list of sequences (lists or arrays)')
```

**Comment by ddsh** [comment]
Yes, the only problem would be the edge case you mentioned before where the first sequence of the list of sequences is empty. 
But this is very degenerate.

**Comment by fchollet** [comment]
But that index error would be caught and then would generate the exception, which would be a reasonable behavior.

**Comment by ddsh** [comment]
Here the exception would catch two different errors: the case where the input is empty and the case where the input is not empty but its first element is empty: [] against [ [] , [array, array] , [array,array,array] ]

**Comment by ddsh** [comment]
Now it makes sense

File: keras/preprocessing/sequence.py [hunk 5]
```diff
@@ -10,17 +10,31 @@ def pad_sequences(sequences, maxlen=None, dtype='int32', padding='pre'):
...
         Support post-padding and pre-padding (default).
     """
     lengths = [len(s) for s in sequences]
-
     nb_samples = len(sequences)
     if maxlen is None:
         maxlen = np.max(lengths)
 
-    x = np.zeros((nb_samples, maxlen)).astype(dtype)
+    array_like = isinstance(sequences[0][0], np.ndarray)
```
Comment (fchollet) [inline keras/preprocessing/sequence.py]: Again, this can trigger an uncaught error.

**Pull request closed without merging**